src/IO/ForkWriteBuffer.cpp

```diff
@@ -0,0 +1,52 @@
+#include <IO/ForkWriteBuffer.h>
+#include <Common/Exception.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_CREATE_IO_BUFFER;
+}
+
+ForkWriteBuffer::ForkWriteBuffer(WriteBufferPtrs && sources_)
+    : WriteBuffer(nullptr, 0), sources(std::move(sources_))
+{
+    if (sources.empty())
+    {
+        throw Exception("Expected non-zero number of buffers for `ForkWriteBuffer`", ErrorCodes::CANNOT_CREATE_IO_BUFFER);
+    }
+    set(sources.front()->buffer().begin(), sources.front()->buffer().size());
+}
+
+
+void ForkWriteBuffer::nextImpl()
+{
+    sources.front()->position() = position();
+
+    try
+    {
+        auto & source_buffer = sources.front();
+        for (auto it = sources.begin() + 1; it != sources.end(); ++it)
+        {
+            auto & buffer = *it;
+            buffer->write(source_buffer->buffer().begin(), source_buffer->offset());
+            buffer->next();
+        }
+        source_buffer->next();
+    }
+    catch (Exception & exception)
+    {
+        exception.addMessage("While writing to ForkWriteBuffer");
+        throw;
+    }
+
+}
+
+ForkWriteBuffer::~ForkWriteBuffer()
+{
+    finalize();
```

Comment:
I see that there is no overloaded finalize method in ForkWriteBuffer, does it make sense to call finalize here?

Reply:
I have called finalize() here so that we call the nextImpl() of ForkWriteBuffer and flush data to the buffers.

Reply:
Probably makes sense to avoid calling `write` for each buffer, but only call `next` (now as default implementation of `finalize` is `finalize() -> finalizeImpl() -> next()`, so `nextImpl` of `ForkWriteBuffer` is called (as Smita mentioned) and we have `write` method also called before `next` for each of the buffers.

Reply:
If we just want to call next method (without write), then I will remove the `finalize()` call from here. The destructor of each WriteBuffer will call the `finalize()->finalizeImpl()->next()`. Please let me know your thoughts.

Reply:
But in general we do not know that all write buffers call `finalize` in destructor (even though it we try to do it everywhere). So if we remove `finalize()` call from here, then for example what if someone passes `ForkWriteBuffer` in buffers list to another `ForkWriteBuffer`, then `next()` will not be called. E.g. we cannot rely that all write buffers indeed call `finalize()->finalizeImpl()->next()` in destructor.